src/main/scala/calc/Tree.scala

```diff
@@ -1,6 +1,30 @@
 package calc
 
-import org.scalajs.core.ir.Position
+import org.scalajs.core.ir.{Position, Types => irtpe}
+
+sealed trait Type {
+  def irtype: irtpe.Type
+}
+
+case class TAny(name: String) extends Type {
+  def irtype = irtpe.AnyType
+}
+
+case object TDouble extends Type {
+  def irtype = irtpe.DoubleType
+}
+
+case class TFun(arity: Int) extends Type {
+  def irtype = irtpe.AnyType
+}
+
+case class TStaticForeignFun(clsName: String, method: String, arity: Int) extends Type {
+  def irtype = irtpe.AnyType
+  override def equals(o: Any) = o match {
+    case that: TStaticForeignFun => this.arity == that.arity
+    case _ => false
+  }
```

Comment:
That `equals`method is suspicious. It might indicate that there is something unelegant about encoding `clsName` and `method` in the *type* of foreign functions.

Anyway, if you redefine `equals`, you must also redefine `hashCode`.